src/iso15118/d20/state/service_detail.cpp

```diff
@@ -44,39 +56,50 @@ message_20::ServiceDetailResponse handle_request(const message_20::ServiceDetail
 
     res.service_parameter_list.clear(); // reset default values
 
+    if (custom_vas_parameters.has_value()) {
+        logf_info("Sending custom vas parameters");
+
+        const auto& vas_services = custom_vas_parameters.value();
+
+        std::vector<uint16_t> parameter_set_ids{};
+        for (auto& vas : vas_services) {
+            parameter_set_ids.push_back(vas.id);
+        }
+        session.offered_services.custom_vas_list[req.service] = parameter_set_ids;
+
+        res.service = req.service;
+        res.service_parameter_list = vas_services;
+        return response_with_code(res, dt::ResponseCode::OK);
+    }
+
     uint8_t id = 0;
 
-    switch (req.service) {
-    case dt::ServiceCategory::DC:
-        res.service = dt::ServiceCategory::DC;
+    if (req.service == message_20::to_underlying_value(dt::ServiceCategory::DC)) {
+        res.service = message_20::to_underlying_value(dt::ServiceCategory::DC);
         for (auto& parameter_set : config.dc_parameter_list) {
             session.offered_services.dc_parameter_list[id] = parameter_set;
             res.service_parameter_list.push_back(dt::ParameterSet(id++, parameter_set));
         }
-        break;
-    case dt::ServiceCategory::DC_BPT:
-        res.service = dt::ServiceCategory::DC_BPT;
+    } else if (req.service == message_20::to_underlying_value(dt::ServiceCategory::DC_BPT)) {
+        res.service = message_20::to_underlying_value(dt::ServiceCategory::DC_BPT);
         for (auto& parameter_set : config.dc_bpt_parameter_list) {
             session.offered_services.dc_bpt_parameter_list[id] = parameter_set;
             res.service_parameter_list.push_back(dt::ParameterSet(id++, parameter_set));
         }
-        break;
-    case dt::ServiceCategory::MCS:
-        res.service = dt::ServiceCategory::MCS;
+    } else if (req.service == message_20::to_underlying_value(dt::ServiceCategory::MCS)) {
+        res.service = message_20::to_underlying_value(dt::ServiceCategory::MCS);
         for (auto& parameter_set : config.mcs_parameter_list) {
             session.offered_services.mcs_parameter_list[id] = parameter_set;
             res.service_parameter_list.push_back(dt::ParameterSet(id++, parameter_set));
         }
-        break;
-    case dt::ServiceCategory::MCS_BPT:
-        res.service = dt::ServiceCategory::MCS_BPT;
+    } else if (req.service == message_20::to_underlying_value(dt::ServiceCategory::MCS_BPT)) {
+        res.service = message_20::to_underlying_value(dt::ServiceCategory::MCS_BPT);
         for (auto& parameter_set : config.mcs_bpt_parameter_list) {
             session.offered_services.mcs_bpt_parameter_list[id] = parameter_set;
             res.service_parameter_list.push_back(dt::ParameterSet(id++, parameter_set));
         }
-        break;
-    case dt::ServiceCategory::Internet:
-        res.service = dt::ServiceCategory::Internet;
+    } else if (req.service == message_20::to_underlying_value(dt::ServiceCategory::Internet)) {
```

Comment:
I think the if clauses for req.service == Internet and req.service == ParkingService are dead code if `m_ctx.feedback.get_vas_parameters()` returns something for these services. In the `feed` function in line 160, the `custom_vas_parameters` is being set to the return value of the vas parameter callback function. If the returned value is not std::nullopt then the if at line 59 in handle_request is true, returning early, thus not reaching this code

Reply:
I forgot to reset `custom_vas_parameters` in the case when internet or parkingstatus got via `m_ctx.feedback.get_vas_parameters()`.
Hopefully it is clearer after my commit.